crates/integrations/datafusion/src/physical_plan/project.rs

```diff
@@ -0,0 +1,661 @@
+
+            let transform_fn = iceberg::transform::create_transform_function(&pf.transform)
+                .map_err(to_datafusion_error)?;
+            let partition_value = transform_fn
+                .transform(source_column)
+                .map_err(to_datafusion_error)?;
+
+            partition_values.push(partition_value);
+        }
+        Ok(partition_values)
+    }
+
+    /// Extract a column by an index path
+    fn extract_column_by_index_path(
+        batch: &RecordBatch,
+        index_path: &[usize],
+    ) -> DFResult<ArrayRef> {
+        if index_path.is_empty() {
+            return Err(DataFusionError::Internal(
+                "Empty index path when extracting partition column".to_string(),
+            ));
+        }
+
+        let mut current_column = batch.column(*index_path.first().unwrap()).clone();
+        for child_index in &index_path[1..] {
+            // We only support traversing nested Structs. Provide explicit errors for unsupported
+            // nested container types to fail early and clearly.
+            let dt = current_column.data_type();
+            match dt {
+                datafusion::arrow::datatypes::DataType::Struct(_) => {
+                    let struct_array = current_column
+                        .as_any()
+                        .downcast_ref::<datafusion::arrow::array::StructArray>()
+                        .ok_or_else(|| {
+                            DataFusionError::Internal(format!(
+                                "Failed to downcast to StructArray while traversing index path {:?} for partition column extraction",
+                                index_path
+                            ))
+                        })?;
+                    current_column = struct_array.column(*child_index).clone();
+                }
+                datafusion::arrow::datatypes::DataType::List(_)
+                | datafusion::arrow::datatypes::DataType::LargeList(_)
+                | datafusion::arrow::datatypes::DataType::FixedSizeList(_, _)
+                | datafusion::arrow::datatypes::DataType::Map(_, _) => {
+                    return Err(DataFusionError::NotImplemented(format!(
+                        "Partitioning on nested list/map types is not supported (encountered {:?}) while traversing index path {:?}",
+                        dt, index_path
+                    )));
+                }
+                other => {
+                    return Err(DataFusionError::Internal(format!(
+                        "Expected struct array while traversing index path {:?} for partition column, got {:?}",
+                        index_path, other
+                    )));
+                }
+            }
+        }
+        Ok(current_column)
+    }
+
+    /// Find the path to a field by its ID (e.g., ["address", "city"]) in the Iceberg schema
+    fn find_field_path(table_schema: &Schema, field_id: i32) -> DFResult<Vec<String>> {
```

Comment:
We might need to consider this function as well @CTTY following our discussion [here](https://github.com/apache/iceberg-rust/pull/1602/files#r2289641476).
It may not be the right place at the moment.